test/test_bmi_model.py

```diff
@@ -41,14 +43,90 @@ def test_bmi_initialize(config: Config) -> None:
-    assert model.optimizer != None
-
+    assert not hasattr(bmi_model, "model")
+
+    bmi_model.initialize(config)
+
+    assert hasattr(bmi_model, "model")
+    assert bmi_model.model != None
+
+    assert bmi_model.learning_rate == config.learning_rate
+    assert bmi_model.optimizer != None
+
+@pytest.mark.parametrize('model', ['bmi_model', 'bmi_model_initialized'])
+def test_bmi_component_name(model, request):
+    m = request.getfixturevalue(model)
+    assert m.get_component_name() == "Tensor Test"
+
+@pytest.mark.parametrize('model', ['bmi_model', 'bmi_model_initialized'])
+def test_bmi_input_item_count(model, request):
+    m = request.getfixturevalue(model)
+    assert m.get_input_item_count() == 1
+
+@pytest.mark.parametrize('model', ['bmi_model', 'bmi_model_initialized'])
+def test_bmi_input_var_names(model, request):
+    m = request.getfixturevalue(model)
+    assert m.get_input_item_count() == 1
+    names = m.get_input_var_names()
+    assert len(names) == 1
+    assert names[0] == "precipitation"
+
+@pytest.mark.parametrize('model', ['bmi_model', 'bmi_model_initialized'])
+def test_bmi_output_item_count(model, request):
+    m = request.getfixturevalue(model)
+    assert m.get_output_item_count() == 1
+
+@pytest.mark.parametrize('model', ['bmi_model', 'bmi_model_initialized'])
+def test_bmi_output_var_names(model, request):
+    m = request.getfixturevalue(model)
+    assert m.get_output_item_count() == 1
+    names = m.get_output_var_names()
+    assert len(names) == 1
+    assert names[0] == "runoff"
+
+@pytest.mark.parametrize("var,expected", [
+                            ("precipitation", 0),
+                            ("runoff", 0)
+                        ])
+def test_bmi_var_grid(bmi_model_initialized, var, expected):
+    m = bmi_model_initialized
+    assert m.get_var_grid(var) == expected
+
+@pytest.mark.parametrize("var", ["var1", "var2"])
+def test_bmi_var_grid_2(bmi_model_initialized, var):
+    m = bmi_model_initialized
+    with pytest.raises(UnknownBMIVariable):
+            m.get_var_grid(var)
+
+@pytest.mark.parametrize("input", [
+                          Tensor([0,0]),
+                          Tensor([0, 1, 2]),
+                          Tensor([2, 1, 0]),
+                          Tensor([ [0] ]),
+                          Tensor([ [0, 1] ]),
+                          Tensor([ [0, 1], [2, 3] ]),
+                          Tensor([ [3, 2], [1, 0] ])
+                        ])
+def test_bmi_get_var_ptr(bmi_model_initialized, input):
+    name = "precipitation"
+    m = bmi_model_initialized
+    m.input = input
+    m._values[name] = m.input
+    data = m.get_value_ptr(name)
+    # Can't test this since data is flattened...
+    #assert data.shape == m.input.shape
+    assert data.shape == m.input.flatten().shape
+    for expected, val in zip( m.input.flatten(), data.flatten() ):
+        assert expected == val
+
+def test_bmi_get_var_ptr_1(bmi_model_initialized, input):
```

Comment:
One thing that is not addressed in the current `get_value_ptr` implementation is how to deal with tensors that have gradients attached. 

This implementation will return a np array of whatever tensor is saved as input, and that's fine if the input tensor doesn't have a gradient attached. However, the following error occurs if the input has a gradient:
```py
>>> m.input = torch.tensor([1.,2.,3.], requires_grad=True)
>>> m.input
tensor([1., 2., 3.], requires_grad=True)
>>> m._values[name] = m.input
>>> m._values[name]
tensor([1., 2., 3.], requires_grad=True)
>>> data = m.get_value_ptr(name)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/taddbindas/projects/bmi_pytorch/bmi_model.py", line 110, in get_value_ptr
    np_array = self._values[name].numpy()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
```

This error would trigger if you are using multiple time steps in your loss calculation where the model's ending state at time `t` relies on timestep `t-1` i.e the inputs contain outputs from the model's previous state. Muskingum-Cunge is a good example of this

Reply:
the solution of this is to use `torch.tensor` everywhere, and only rely on np arrays when reading from disk, or operations that are not directly tied with `nn.Layer()` inputs of outputs.

However, this breaks the type hinting encouraged from `bmi-python` as it only wants numpy ndarrays.

Reply:
That is exactly the problem I want to work on here!  Just needed to get the foundation in place to see what works, what doesn't, and what the limitations are.  I think there is a path forward for re-connecting the gradients, but its going to take a couple "bmi workarounds" 😄